slang-checks/src/main/java/com/sonarsource/slang/checks/DuplicatedFunctionImplementationCheck.java

```diff
@@ -47,12 +51,17 @@
   @Override
   public void initialize(InitContext init) {
     init.register(TopLevelTree.class, (ctx, tree) -> {
-      List<FunctionDeclarationTree> functionDeclarations = tree.descendants()
-        .filter(FunctionDeclarationTree.class::isInstance)
-        .map(FunctionDeclarationTree.class::cast)
-        .collect(Collectors.toList());
+      Map<Tree, List<FunctionDeclarationTree>> functionsByParents = new HashMap<>();
```

Comment:
A different implementation could be to make the check register on `Tree.class` and check its children. I'm not sure whether it would be better.

Reply:
Yes, I thought about this as well. I wasn't sure which to go with. In the end I chose this one to avoid using `instanceof` in the rule.